### PR TITLE
user deactivate flow

### PR DESCRIPTION
**Pull request by adithyadinesh0412**

### CODE REVIEW

File: src/services/helper/account.js [hunk 1]
```diff
@@ -164,7 +164,7 @@ module.exports = class AccountHelper {
 			}
 
 			await userQueries.updateUser({ id: user.id }, update)
-			await utilsHelper.redisDel(email)
+			// await utilsHelper.redisDel(email)
```
Comment (rakeshSgr) [inline src/services/helper/account.js:167]: @adithyadinesh0412 is this line required?

File: src/services/helper/account.js [hunk 2]
```diff
@@ -982,4 +982,125 @@ module.exports = class AccountHelper {
...
+	 * @name deactivateUser
+	 * @param {string} bodyData.email - email of user.
+	 * @returns {JSON} change role success response
+	 */
+	static async deactivate(bodyData) {
+		const projection = []
+		const today = new Date()
+		const userRedisData = await utilsHelper.redisGet(bodyData.email.toLowerCase())
+		try {
+			if (userRedisData && userRedisData.action == 'deactivateUser' && userRedisData.otp == bodyData.otp) {
```
Comment (rakeshSgr) [inline src/services/helper/account.js:998]: @adithyadinesh0412  action should come from common config  'deactivateUser'

File: src/services/helper/account.js [hunk 3]
```diff
@@ -982,4 +982,125 @@ module.exports = class AccountHelper {
...
+				return common.failureResponse({
+					message: 'USER_DOESNOT_EXISTS',
+					statusCode: httpStatusCode.bad_request,
+					responseCode: 'CLIENT_ERROR',
+				})
+			}
+
+			const userData = await utilsHelper.redisGet(bodyData.email.toLowerCase())
+
+			if (userData && userData.action === 'deactivateUser') {
```
Comment (rakeshSgr) [inline src/services/helper/account.js:1072]: @adithyadinesh0412 get action 'deactivateUser' from common config

File: src/services/helper/account.js [hunk 4]
```diff
@@ -982,4 +982,125 @@ module.exports = class AccountHelper {
...
+					verify: bodyData.email.toLowerCase(),
+					action: 'deactivateUser',
+					otp,
+				}
+				const res = await utilsHelper.redisSet(
+					bodyData.email.toLowerCase(),
+					redisData,
+					common.otpExpirationTime
+				)
+				if (res !== 'OK') {
```
Comment (rakeshSgr) [inline src/services/helper/account.js:1089]: @adithyadinesh0412 get response code from  common response code

File: src/controllers/v1/account.js [hunk 5]
```diff
@@ -235,4 +235,32 @@ module.exports = class Account {
...
+		const params = req.body
+		try {
+			const result = await accountHelper.deactivate(params)
+			return result
+		} catch (error) {
+			return error
+		}
+	}
+
+	async generateDisableOtp(req) {
```
Comment (rakeshSgr) [inline src/controllers/v1/account.js:257]: @adithyadinesh0412 method details are missing

Comment (rakeshSgr) [reply src/controllers/v1/account.js:266]: @adithyadinesh0412 there is an existing generateOtp function can we use that function ?

File: src/services/helper/account.js [hunk 6]
```diff
@@ -982,4 +982,125 @@ module.exports = class AccountHelper {
...
+
+			const userData = await utilsHelper.redisGet(bodyData.email.toLowerCase())
+
+			if (userData && userData.action === 'deactivateUser') {
+				otp = userData.otp // If valid then get previuosly generated otp
+			} else {
+				isValidOtpExist = false
+			}
+			if (!isValidOtpExist) {
+				otp = Math.floor(Math.random() * 900000 + 100000) // 6 digit otp
```
[the remark this answers is not in the document]
Comment (rakeshSgr) [reply src/services/helper/account.js:1102]: @adithyadinesh0412 please do the generic method for generating the otp

File: src/database/migrations/20230531105847-create-users-table.js [hunk 7]
```diff
@@ -32,6 +32,11 @@ module.exports = {
 				allowNull: false,
 				defaultValue: 'ACTIVE',
 			},
+			status_updated_at: {
```
Comment (ankitpws) [inline src/database/migrations/20230531105847-create-users-table.js:35]: create new migration for this changes @adithyadinesh0412

File: src/migrations/20220722162957-notificationTemplates.js [hunk 8]
```diff
@@ -14,6 +14,11 @@ let emailTemplates = [
 		subject: 'Your OTP to sign-up on MentorED',
 		body: '<div><p>Dear {name},</p> Your OTP to complete the registration process is {otp}. Please enter the OTP to complete the registration. For your security, please do not share this OTP with anyone.</div>',
 	},
+	{
```
Comment (ankitpws) [inline src/migrations/20220722162957-notificationTemplates.js:17]: dont update in migration

File: src/.env.sample [hunk 9]
```diff
@@ -140,3 +140,12 @@ INVITEE_EMAIL_TEMPLATE_CODE=invite_user
 
 #Email template for upload status
 ADMIN_INVITEE_UPLOAD_EMAIL_TEMPLATE_CODE=invitee_upload_status
+
+#Email template for registration
+REGISTRATION_EMAIL_TEMPLATE_CODE=registration
```
Comment (ankitpws) [inline src/.env.sample:145]: Why additional ENV variables added and if added then templets are missing

**Review by ankitpws — CHANGES_REQUESTED** [review]
Reviewed on 9 Nov 3

File: src/.env.sample (hunk 9 above)
Comment (adithyadinesh0412) [reply src/.env.sample:145]: template also added in seeders. @ankitpws

File: src/controllers/v1/account.js (hunk 5 above)
Comment (adithyadinesh0412) [reply src/controllers/v1/account.js:266]: added a generic function in generics/utils

File: src/generics/utils.js [hunk 10]
```diff
@@ -329,6 +329,15 @@ function removeParentEntityTypes(data) {
 	return data.filter((item) => !parentIds.includes(item.id))
 }
 
+const generateOtp = (otpLength, email, action) => {
+	// Generate a random 6-digit number
```
Comment (ankitpws) [inline src/generics/utils.js:333]: generate OTP can be normal function why email ID is needed @adithyadinesh0412

File: src/services/account.js [hunk 11]
```diff
@@ -1056,4 +1056,179 @@ module.exports = class AccountHelper {
...
+
+			// check if the user have an OTP for deactivate functionality
+			if (userData && userData.action === common.deactivateAction) {
+				otp = userData.otp // If valid then get previuosly generated otp
+			} else {
+				isValidOtpExist = false // if no data found in redis , set this flag flase to generate new OTP
+			}
+
+			if (!isValidOtpExist) {
+				const redisData = utils.generateOtp(6, bodyData.email.toLowerCase(), common.deactivateAction)
```
Comment (ankitpws) [inline src/services/account.js:1180]: get OTP from utils and prepare data here @adithyadinesh0412

**Review by ankitpws — CHANGES_REQUESTED** [review]
Reviewed ON 14 Nov

**Review by ankitpws — APPROVED** [review]
Approved ON 14 Nov 5:20PM

**Pull request closed without merging**